samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java

```diff
@@ -113,21 +146,50 @@
    */
   public ClusterBasedJobCoordinator(Config coordinatorSystemConfig) {
 
-    MetricsRegistryMap registry = new MetricsRegistryMap();
+    this.metrics = new MetricsRegistryMap();
 
     //build a JobModelReader and perform partition assignments.
-    jobModelManager = buildJobModelManager(coordinatorSystemConfig, registry);
+    jobModelManager = buildJobModelManager(coordinatorSystemConfig, this.metrics);
     config = jobModelManager.jobModel().getConfig();
+    hasDurableStores = new StorageConfig(config).hasDurableStores();
     state = new SamzaApplicationState(jobModelManager);
+    this.partitionMonitor = getPartitionCountMonitor(config);
     clusterManagerConfig = new ClusterManagerConfig(config);
     isJmxEnabled = clusterManagerConfig.getJmxEnabled();
 
     jobCoordinatorSleepInterval = clusterManagerConfig.getJobCoordinatorSleepInterval();
 
     // build a container process Manager
-    containerProcessManager = new ContainerProcessManager(config, state, registry);
+    containerProcessManager = new ContainerProcessManager(config, state, this.metrics);
   }
 
+  private StreamPartitionCountMonitor getPartitionCountMonitor(Config config) {
+    Map<String, SystemAdmin> systemAdmins = new HashMap<>();
+    JavaSystemConfig sysConfig = new JavaSystemConfig(config);
+    sysConfig.getSystemFactories().entrySet().forEach(entry -> {
+        systemAdmins.put(entry.getKey(), entry.getValue().getAdmin(entry.getKey(), config));
+      });
+    StreamMetadataCache streamMetadata = new StreamMetadataCache(Util.javaMapAsScalaMap(systemAdmins), 0, SystemClock.instance());
+    Set<SystemStream> inputStreamsToMonitor = new TaskConfigJava(config).getAllInputStreams();
+    try {
+      checkArgument(!inputStreamsToMonitor.isEmpty(), "Input streams can not be empty for a job.");
+    } catch (Exception e) {
+      throw new SamzaException("Input streams to a job can not be empty.");
+    }
+
+    return new StreamPartitionCountMonitor(
+        inputStreamsToMonitor,
+        streamMetadata,
+        this.metrics,
+        new JobConfig(config).getMonitorPartitionChangeFrequency(),
+        streamsChanged -> {
+        if (this.hasDurableStores) {
+          log.error("Input topic partition count changed in a job with durable state. Failing the job.");
+          this.state.status.set(SamzaApplicationState.SamzaAppStatus.FAILED);
+        }
+        coordinatorException = new PartitionChangeException("Input topic partition count changes detected.");
```

Comment:
Instead of setting an exception here, would it be clearer to mark a boolean that `partitionChanged= true` and have the JC poll for it, and throw an Exception in its loop? - The intent is more obvious that way.

Reply:
The concern that I have is to introduce one variable for each possible failure/restart scenario that we may add in the future in the JC. Hence, using an exception and setting it with proper type of exception in this particular case.

Reply:
Good point, makes sense!